src/AnalysisEpic.cxx

```diff
@@ -0,0 +1,424 @@
+
+  // calculate Q2 weights
+  CalculateEventQ2Weights();
+
+  // counters
+  Long64_t numNoBeam, numEle, numNoEle, numNoHadrons, numProxMatched;
+  numNoBeam = numEle = numNoEle = numNoHadrons = numProxMatched = 0;
+
+  
+
+  cout << "begin event loop..." << endl;
+  tr.SetEntriesRange(1,maxEvents);
+  do{
+    if(tr.GetCurrentEntry()%10000==0) cout << tr.GetCurrentEntry() << " events..." << endl;
+
+    // resets
+    kin->ResetHFS();
+    kinTrue->ResetHFS();
+
+    
+    double maxP = 0;
+    int genEleID = -1;
+    bool foundBeamElectron = false;
+    bool foundBeamIon = false;
+
+    // Index maps for particle sets
+    std::map <double,int> genidmap; // <pz, index_gen>
+    std::map <int,int> mcidmap; // <index_mc, index_gen>
+    std::map <int,int> trackidmap; // <index, index_mc>  
+    std::map <int,int> trackstatmap; // <index, genstatus>
+
+    // Particles vectors
+    // The index of the vectors correspond to their for loop idx
+    std::vector<Particles> genpart;    // mcID --> igen
+    std::vector<Particles> mcpart;     // mcID --> imc
+    std::vector<Particles> trackpart;  // mcID --> (imc of matching mcpart) or (-1 if no match is found)
+
+    /*
+      GenParticles loop
+    */
+
+    for(int igen=0; igen<hepmcp_PDG.GetSize(); igen++) {
+
+      int pid_ = hepmcp_PDG[igen];
+
+      double px_ = hepmcp_psx[igen];
+      double py_ = hepmcp_psy[igen];
+      double pz_ = hepmcp_psz[igen];
+      double e_  = hepmcp_E[igen];
+     
+      double p_ = sqrt(pow(hepmcp_psx[igen],2) + pow(hepmcp_psy[igen],2) + pow(hepmcp_psz[igen],2));
+      double mass_ = (fabs(pid_)==211)?constants::pimass:(fabs(pid_)==321)?constants::kmass:(fabs(pid_)==11)?constants::emass:(fabs(pid_)==13)?constants::mumass:(fabs(pid_)==2212)?constants::pmass:0.;
+
+      // Add to genpart
+      Particles part;
+
+      part.pid=pid_;
+      part.charge = (pid_ == 211 || pid_ == 321 || pid_ == 2212 || pid_ == -11 || pid_ == -13)?1:(pid_ == -211 || pid_ == -321 || pid_ == -2212 || pid_ == 11 || pid_ == 13)?-1:0;
+      part.mcID=igen;
+      part.vecPart.SetPxPyPzE(px_,py_,pz_,e_);
+      genpart.push_back(part);
+      
+      genidmap.insert({pz_,igen});
+
+    }
+
+    /*
+      MCParticles loop
+    */
+
+    for(int imc=0; imc < mcpart_PDG.GetSize(); imc++){
+
+      int pid_ = mcpart_PDG[imc];
+      double px_ = mcpart_psx[imc];
+      double py_ = mcpart_psy[imc];
+      double pz_ = mcpart_psz[imc];
+      double m_ = mcpart_m[imc];
+      double e_ = sqrt(px_*px_+py_*py_+pz_*pz_+m_*m_);
+
+      // Add to mcpart
+      Particles part;
+
+      part.pid=pid_;
+      part.charge = (pid_ == 211 || pid_ == 321 || pid_ == 2212 || pid_ == -11 || pid_ == -13)?1:(pid_ == -211 || pid_ == -321 || pid_ == -2212 || pid_ == 11 || pid_ == 13)?-1:0;
+      part.mcID=imc;
+      part.vecPart.SetPxPyPzE(px_,py_,pz_,e_);
+      mcpart.push_back(part);
+      
+      int igen=-1;
+      if(auto search = genidmap.find(pz_); search != genidmap.end())
```

Comment:
how accurate is `std::map::find` for `double` keys?

Reply:
After scanning through StackOverflow (https://stackoverflow.com/questions/6684573/floating-point-keys-in-stdmap) it seems we may be in the clear if we are certain the "pz" in GenParticles is explicitly the same value as the "pz" in MCParticles, decimal precision and all. What do you recommend as a better approach in this case? If we do something like closest element, we run the risk of incorrectly matching particles that appear in GenParticles but not in MCParticles, or vice versa.

Reply:
Could you loop through `genpart` instead, and look for the closest match?

Reply:
https://github.com/eic/sidis-eic/issues/219